cosmoz-grouped-list-templatize-behavior.js

```diff
@@ -0,0 +1,300 @@
+		return {
+			[this.as]: true,
+			[this.indexAs]: true,
+			folded: true,
+			expanded: true,
+			selected: true,
+			highlighted: true
+		};
+	}
+
+	/**
+	 * Set template defaults.
+	 * @param {array} addedNodes Added nodes.
+	 * @returns {void}
+	 */
+	_onTemplatesChange({addedNodes}) {
+		const ctors = this._ctors;
+
+		if (ctors && Object.keys(ctors) > 0) {
+			return;
+		}
+
+		const templates = Array.from(addedNodes)
+			.filter(n => n.matches && n.matches('template[data-type]'));
+
+		if (templates.length === 0) {
+			console.warn('cosmoz-grouped-list requires templates');
+			return;
+		}
+
+		this._ctors = templates.reduce((ctors, template) => {
+			const type = template.dataset.type;
+			ctors[type] = templatize(template, this, {
+				instanceProps: this._getInstanceDefaultProps(),
+				parentModel: true,
+				forwardHostProp: this._forwardHostProp,
+				notifyInstanceProp: this._notifyInstanceProp
+			});
+			return ctors;
+		}, {});
+	}
+
+	/**
+	 * Reuse an existing TemplateInstance or create a new one if there is not any available.
+	 *
+	 * @param		{String}						type					 The type of instance to return
+	 * @param		{Object}						props					 Properties to set on the instance
+	 * @param		{TemplateInstance}	prevInstance	 The previously used instance
+	 * @param		{Boolean}						flush					 True if instance properties should be flushed.
+	 * @returns {TemplateInstance}								 The instance to use
+	*/
+	_getInstance(type, props = {}, prevInstance, flush = true) {
+		const ctors = this._ctors;
+		if (ctors == null || Object.keys(ctors).length === 0) {
+			console.warn('cosmoz-grouped-list templates are required.');
+			return;
+		}
+		if (ctors[type] == null) {
+			console.warn(`cosmoz-grouped-list template for ${type} type not found.`);
+		}
+
+		if (prevInstance && prevInstance.__type === type) {
+			this._forwardProperties(prevInstance, props, flush);
+			return prevInstance;
+		}
+
+		if (prevInstance) {
+			this._reuseInstance(prevInstance);
+		}
+
+		const {_reusableInstances: reusable, _instances: instances} = this;
+
+		let instance = reusable.find(({__type}) => __type === type);
+
+		if (instance) {
+			reusable.splice(reusable.indexOf(instance), 1);
+			this._forwardProperties(instance, props, flush);
+		} else {
+			instance = new ctors[type](props);
+			instance.__type = type;
+			instance.element = instance.root.querySelector('*');
+		}
+		instances.push(instance);
+		return instance;
+	}
+
+	/**
+	 * Get an instance by a matching property value.
+	 * @param {string} prop Property
+	 * @param {any} value Value.
+	 * @return {object} Matching instance.
+	 */
+	_getInstanceByProperty(prop, value) {
+		return this._instances
+			.find(instance => this._getInstanceProperty(instance, prop) === value);
+	}
+
+	/**
+	 * Get an instance property.
+	 * @param {object} instance Instance.
+	 * @param {string} prop Property in instance.
+	 * @return {any} Instance property.
+	 */
+	_getInstanceProperty(instance, prop) {
```

Comment:
I think it's used this way because `_getInstanceProperty` is called from outside (not sure about this; if this is true, it should be refactored, because it's a _private method) and this avoid duplicating code.